cocos/core/renderer/scene/render-scene.ts

```diff
@@ -165,13 +166,36 @@ export class RenderScene {
     }
 
     public setMainLight (dl: DirectionalLight) {
-        dl.attachToScene(this);
         this._mainLight = dl;
     }
 
     public unsetMainLight (dl: DirectionalLight) {
-        dl.detachFromScene();
-        this._mainLight = null;
+        if (this._mainLight === dl) {
+            const dlList = this._directionalLights;
+            if (dlList.length) {
+                this._mainLight = dlList[dlList.length - 1];
+                if (this._mainLight.node) { // trigger update
+                    this._mainLight.node.setRotation(this._mainLight.node.rotation);
```

Comment:
In the future if we check equality of set value in `setRotation`, the trigger will fail, how it's triggered exactly ? By TRANSFORM_CHANGED event ? if so, you can emit a event here direclty.